Allow specifying index in pinned queries

davishmcclurg opened the pull request:
The current `ids` option doesn't allow pinning a specific document in a
single index when searching over multiple indices. This introduces a
`documents` option, which is an array of `_id` and optional `_index`
fields to allow index-specific pins.

Closes: https://github.com/elastic/elasticsearch/issues/67855

I copied some of the `Item` code for the `MoreLikeThisQueryBuilder`: https://github.com/elastic/elasticsearch/blob/0700359fc2ff62a66ee5d805e85f9007f9277f8e/server/src/main/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilder.java#L138

and simplified it a bit. ~I haven't added any tests yet—I wanted to get
some feedback on the approach before spending more time on it.~ I'm not
a Java expert by any means, so let me know if I've committed any faux pas.

davishmcclurg committed:
Allow specifying index in pinned queries

The current `ids` option doesn't allow pinning a specific document in a
single index when searching over multiple indices. This introduces a
`documents` option, which is an array of `_id` and optional `_index`
fields to allow index-specific pins.

Closes: https://github.com/elastic/elasticsearch/issues/67855

davishmcclurg committed:
Fix line length issue

davishmcclurg commented:
Here's an example and comparison between `ids` and `documents`:

```
% curl 'localhost:9200/test1/_doc/a' -H 'Content-Type: application/json' -d '{}'

{"_index":"test1","_id":"a","_version":1,"result":"created","_shards":{"total":2,"successful":1,"failed":0},"_seq_no":0,"_primary_term":1}%

% curl 'localhost:9200/test1/_doc/b' -H 'Content-Type: application/json' -d '{}'

{"_index":"test1","_id":"b","_version":1,"result":"created","_shards":{"total":2,"successful":1,"failed":0},"_seq_no":1,"_primary_term":1}%

% curl 'localhost:9200/test2/_doc/a' -H 'Content-Type: application/json' -d '{}'

{"_index":"test2","_id":"a","_version":1,"result":"created","_shards":{"total":2,"successful":1,"failed":0},"_seq_no":0,"_primary_term":1}%

% curl 'localhost:9200/test2/_doc/b' -H 'Content-Type: application/json' -d '{}'

{"_index":"test2","_id":"b","_version":1,"result":"created","_shards":{"total":2,"successful":1,"failed":0},"_seq_no":1,"_primary_term":1}%

% curl -s 'localhost:9200/test*/_search' -H 'Content-Type: application/json' -d '{
  "query": {
    "pinned": {
      "organic": {
        "match_all": {}
      },
      "ids": ["a"]
    }
  }
}' | jq '.hits.hits'

[
  {
    "_index": "test1",
    "_id": "a",
    "_score": 1.7014122e+38,
    "_source": {}
  },
  {
    "_index": "test2",
    "_id": "a",
    "_score": 1.7014122e+38,
    "_source": {}
  },
  {
    "_index": "test1",
    "_id": "b",
    "_score": 1,
    "_source": {}
  },
  {
    "_index": "test2",
    "_id": "b",
    "_score": 1,
    "_source": {}
  }
]

% curl -s 'localhost:9200/test*/_search' -H 'Content-Type: application/json' -d '{
  "query": {
    "pinned": {
      "organic": {
        "match_all": {}
      },
      "documents": [
        {
          "_index": "test1",
          "_id": "a"
        }
      ]
    }
  }
}' | jq '.hits.hits'

[
  {
    "_index": "test1",
    "_id": "a",
    "_score": 1.7014122e+38,
    "_source": {}
  },
  {
    "_index": "test1",
    "_id": "b",
    "_score": 1,
    "_source": {}
  },
  {
    "_index": "test2",
    "_id": "a",
    "_score": 1,
    "_source": {}
  },
  {
    "_index": "test2",
    "_id": "b",
    "_score": 1,
    "_source": {}
  }
]
```

davishmcclurg committed:
Switch to dis max for pinned queries

With `documents`, multiple queries can match the same document:

```json
[
  {
    "_index": "test1",
    "_id": "a"
  },
  {
    "_id": "a"
  }
]
```

Combining the scores with `should` is causing issues:

```json
{
  "type": "illegal_argument_exception",
  "reason": "boost must be a positive float, got Infinity"
}
```

This switches to dis max to use the highest score instead.

davishmcclurg committed:
Add `toString` method for items

This is used in duplicate document exceptions.

davishmcclurg committed:
Use `orElseGet` to get query items

This should work since `ids` and `documents` cannot both be set.
`

davishmcclurg committed:
Register named writeable for pinned documents

davishmcclurg committed:
Clean up with checkstyle and code formatter

davishmcclurg committed:
Fix another line length issue

davishmcclurg committed:
Stop using Optional for fields

IntelliJ is complaining about this and it does seem like it's making
things more complicated for no benefit.

davishmcclurg committed:
Use plain `Writeable` for pinned documents

Not sure why I thought `NamedWriteable` was necessary before. Switched
now because I couldn't get the registry working in tests and this seems
simpler.

davishmcclurg committed:
Merge branch 'master' into pinned-index

davishmcclurg committed:
Add tests for pinned documents

This mostly just copies the existing pinned ids tests and modifies them
to use items instead. The main new test is `testMultiIndexDocuments`
which tests pinning the same id in multiple indices.

davishmcclurg commented:
I added some tests, which I mostly copied from the existing ones.

Sorry for the noisy commits—I opened this earlier than I should have. Let me know if I should rebase.

elasticmachine commented:
Pinging @elastic/es-search (Team:Search)

jtibshirani commented on x-pack/plugin/search-business-rules/src/main/java/org/elasticsearch/xpack/searchbusinessrules/PinnedQueryBuilder.java:
```diff
@@ -49,56 +58,184 @@
     public static final int MAX_NUM_PINNED_HITS = 100;
 
     private static final ParseField IDS_FIELD = new ParseField("ids");
+    private static final ParseField DOCUMENTS_FIELD = new ParseField("documents");
```
I think it'd be nice to use "docs" instead of "documents" to match our multi-get and multi-term vectors requests.

jtibshirani commented on x-pack/plugin/search-business-rules/src/main/java/org/elasticsearch/xpack/searchbusinessrules/PinnedQueryBuilder.java:
```diff
@@ -176,32 +334,39 @@ protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws
             // Use dismax to always take the larger (ie pinned) of the organic vs pinned scores
             List<Query> organicAndPinned = new ArrayList<>();
-            organicAndPinned.add(pinnedQueries.build());
+            organicAndPinned.add(new DisjunctionMaxQuery(pinnedQueries, 0));
```
Is there a motivation for switching from boolean to dismax query? I think the scoring should be identical between the two.

davishmcclurg replied:
Yeah this got a bit lost in the messy commits: 7d98d38ed853bf05db2b0d4a117c8f586c0b7b2e

I'm not entirely sure, but it seemed like `should` was combining scores in a way that caused issues when multiple `documents` clauses matched the same document.

jtibshirani replied:
Oh I see, I didn't realize we now allow duplicate IDs in the list. I wonder if we should just require `_index` to always be set, to keep the behavior as simple as possible. I can't think of a use case for mixing both ID-only with ID-plus-index items -- it seems like you'd always want one or the other?

davishmcclurg replied:
I'm not sure of a use case either. I was thinking someone might want a specific document (`_index` and `_id`) as the top result and then documents with the same id in other indices after that. If that doesn't seem helpful, requiring `_index` sounds ok to me too.

jtibshirani commented on x-pack/plugin/search-business-rules/src/main/java/org/elasticsearch/xpack/searchbusinessrules/PinnedQueryBuilder.java:
```diff
@@ -176,32 +334,39 @@ protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws
                 assert pinScore < lastScore;
                 lastScore = pinScore;
                 boostNum--;
+                BooleanQuery.Builder documentQueries = new BooleanQuery.Builder();
```
Since this method is called on a single index shard, it feels simplest to check check the index right here through `SearchExecutionContext#indexMatches` and omit the item if it doesn't match.

davishmcclurg replied:
Nice—that makes things much simpler. Do I need to handle the case where `pinnedQueries` would be empty or is creating an empty `DisjunctionMaxQuery` ok?

jtibshirani replied:
I think it's fine to have an empty `DisjunctionMaxQuery`, although it doesn't explicitly optimize this case. Depending on what we decide with allowing duplicates (by making `_index` required or not), we may be also able to stick with a `BooleanQuery` here.

jtibshirani commented on x-pack/plugin/search-business-rules/src/main/java/org/elasticsearch/xpack/searchbusinessrules/PinnedQueryBuilder.java:
```diff
@@ -49,56 +58,184 @@
     protected void doWriteTo(StreamOutput out) throws IOException {
-        out.writeStringCollection(this.ids);
+        out.writeOptionalStringCollection(this.ids);
+        if (documents == null) {
```
We'll need to add a check to make sure we don't send 'documents' to nodes on older versions, since they aren't expecting this key. We can use `out.getVersion().onOrAfter(...)` to check the version. The input side using `StreamInput` will need a similar check, since old nodes will send data in the old format.

davishmcclurg replied:
Ah, that makes sense. How is the version number handled when backporting? I used `getVersion().before(Version.V_8_0_0)` to gate the changes, but I assume `Version.V_8_0_0` changes if we backport?

jtibshirani replied:
Right, when backporting we'll switch it to something like `getVersion().before(Version.V_7_15_0)`, then once that's merged we'll immediately change this on master to be `Version.V_7_15_0` too. I'm also happy to handle the backport since it can be tricky.

jtibshirani reviewed (commented):
Thanks @davishmcclurg for the PR, it looks great overall!

From a high-level, I wanted to check we're happy with the API. Instead of adding a new `documents` section, I guess we could have extended `ids` to allow object values like `{"_id": ..., "_index": ...}`.  I like your approach better, since it's easy to prevent duplicate items in the list. Plus `documents` is a more natural name for this data. Tagging @markharwood in case he has thoughts.

Update: I didn't realize we allowed duplicate IDs in the list. I'll follow-up about this in another comment. The `documents` approach still feels most natural to me.

davishmcclurg commented on x-pack/plugin/search-business-rules/src/internalClusterTest/java/org/elasticsearch/xpack/searchbusinessrules/PinnedQueryBuilderIT.java:
```diff
@@ -55,6 +58,16 @@ public void testIdInsertionOrderRetained() {
     }
 
+    public void testDocumentInsertionOrderRetained() {
+        Item[] items = randomArray(10, Item[]::new, () -> new Item(randomAlphaOfLength(64), randomAlphaOfLength(256)));
```
Oh yeah, good call. Should I move `testIdInsertionOrderRetained` as well?

jtibshirani replied:
Yep let's move that too (oops, guess that was in the wrong place before).

davishmcclurg committed:
Review feedback

- Rename `documents` to `docs` to be more consistent with other APIs.
- Use `Collections.emptyList()` for creating empty list.
- Handle old versions in stream input/output.
- Use `indexMatches` to skip items that aren't relevant to the context.

davishmcclurg committed:
Merge branch 'master' into pinned-index

davishmcclurg committed:
Merge branch 'master' into pinned-index

davishmcclurg committed:
Move unit tests out of integration tests

davishmcclurg commented:
Do I need to update [`org.elasticsearch.xpack.core.index.query.PinnedQueryBuilder`](https://github.com/elastic/elasticsearch/blob/063a1f2d30d041f14e982f50bfdbdb6251123b9a/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/index/query/PinnedQueryBuilder.java#L28) as well? And will that need the same version checks around `doWriteTo`? https://github.com/elastic/elasticsearch/blob/063a1f2d30d041f14e982f50bfdbdb6251123b9a/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/index/query/PinnedQueryBuilder.java#L58-L61

jtibshirani commented:
@davishmcclurg sorry for the delay. I checked with the team and we're happy with this API:
* Add a new `docs` section that accepts a list of `(_index, _id)` pairs. Both index and ID must be specified.
* Only one of `ids` and `docs` can be specified, to avoid the complexity of duplicate IDs and because there's not a strong use case for mixing them.

> Do I need to update [`org.elasticsearch.xpack.core.index.query.PinnedQueryBuilder`](https://github.com/elastic/elasticsearch/blob/063a1f2d30d041f14e982f50bfdbdb6251123b9a/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/index/query/PinnedQueryBuilder.java#L28) as well?

Good catch -- this class belongs to the Java High-level REST Client. This is being phased out by a new, auto-generated client, but it'd be good to update it in this PR to keep it in sync. I don't think we need to account for input/ output versions in this client class.

davishmcclurg commented:
@jtibshirani no worries—thanks for checking with the team!

> * Both index and ID must be specified.

I'm still a little unsure on this point. It looks like using `SearchExecutionContext#indexMatches` would allow you to use a wildcard (`"_index": "*"`), which I think would be the same as not requiring it at all. Wildcards also mean we need to handle the issue of duplicate IDs in multiple indices (switching to dis max).

jtibshirani commented:
This is a good point. Even if we disallowed wildcard expressions, we could still run into duplicates because of index aliases. For example, there could be two aliases `alias1` and `alias2` that refer to the same concrete index -- searching both these aliases with the same document ID would produce a duplicate result. For context, `term` queries on the `_index` field also accept wildcard expressions (kind of unfortunate), so it wouldn't help to switch to them. Here's a new proposal:
* Still require `docs` to include both index and ID, to keep the API as straightforward as possible
* For simplicity also reject wildcard expressions by checking `Regex#isSimpleMatchPattern`
* Make sure to use dismax to handle potential duplicates, as you currently do

I also remembered we should update the docs in `pinned-query.asciidoc` to give an example of the new parameter. Thanks for all the iterations, this turned out to be trickier than I realized at first!

davishmcclurg committed:
Merge branch 'master' into pinned-index

davishmcclurg committed:
Make `_index` required for pinned docs

We haven't come up with a good use case for pinning a doc across
multiple indices and this keeps the API as simple as possible.

davishmcclurg committed:
Reject wildcard `_index` in pinned docs

Trying to keep things simple for now.

davishmcclurg committed:
Update pinned query docs for `docs`

davishmcclurg committed:
Add `docs` to pinned query in HLRC

davishmcclurg commented:
> * Still require `docs` to include both index and ID, to keep the API as straightforward as possible
> * For simplicity also reject wildcard expressions by checking `Regex#isSimpleMatchPattern`
> * Make sure to use dismax to handle potential duplicates, as you currently do
> 
> I also remembered we should update the docs in `pinned-query.asciidoc` to give an example of the new parameter.

Sounds good :+1: I just pushed a few commits that I think cover everything.

jtibshirani commented on docs/reference/query-dsl/pinned-query.asciidoc:
```diff
@@ -31,6 +31,51 @@ GET /_search
+          "_id": "100"
+        }
+      ],
+      "organic": {
```
Small comment, I think it's a bit clearer to use the same `match` query here as above.

jtibshirani commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/index/query/PinnedQueryBuilder.java:
```diff
@@ -29,34 +32,95 @@
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.startObject();
+            builder.field(INDEX_FIELD.getPreferredName(), this.index);
```
Small comment, we usually don't use the `this` reference unless it helps avoid ambiguity (like in a constructor).

davishmcclurg replied:
That makes sense. It wasn't clear to me when `this` was necessary. I removed it here but left it in other spots where it was already used.

jtibshirani commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/index/query/PinnedQueryBuilder.java:
```diff
@@ -29,34 +32,95 @@
         return NAME;
     }
 
+    public PinnedQueryBuilder(QueryBuilder organicQuery) {
```
We can avoid adding this extra constructor, it doesn't seem helpful to create a pinned query with no pinned documents?

davishmcclurg replied:
Good call—I think I added this to keep a test working. I updated the test instead :+1:

jtibshirani commented on x-pack/plugin/search-business-rules/src/internalClusterTest/java/org/elasticsearch/xpack/searchbusinessrules/PinnedQueryBuilderIT.java:
```diff
@@ -222,5 +233,42 @@ public void testHighlight() throws Exception {
         assertThat(highlight.fragments()[0].toString(), equalTo("<em>the</em> <em>quick</em> <em>brown</em> fox"));
     }
+
+    public void testMultiIndexDocs() throws Exception {
```
It'd be great to add a test for the tricky case you uncovered around index aliases, where the same document could match multiple pinned entries.

davishmcclurg replied:
Added here: addc260d96204aaf481f0435e9426aadae679ff0

jtibshirani commented on x-pack/plugin/search-business-rules/src/main/java/org/elasticsearch/xpack/searchbusinessrules/PinnedQueryBuilder.java:
```diff
@@ -49,56 +56,202 @@
+        }
+    }
+
+    public PinnedQueryBuilder(QueryBuilder organicQuery) {
```
Same comment as above about this constructor, I think we can remove it.

jtibshirani commented on x-pack/plugin/search-business-rules/src/main/java/org/elasticsearch/xpack/searchbusinessrules/PinnedQueryBuilder.java:
```diff
@@ -113,21 +266,39 @@ public QueryBuilder organicQuery() {
      * Returns the pinned ids for the query.
      */
     public List<String> ids() {
+        if (this.ids == null) return Collections.emptyList();
```
Small comment, we try to avoid single-line if statements.

jtibshirani reviewed (commented):
This is looking good! I left a few small comments, but plan to merge after they're addressed.

I'll handle the backport since handling our backwards compatibility tests can be a bit tricky. I'll also add a REST test before merging, which help us catch any bwc issues.

davishmcclurg committed:
Merge branch 'master' into pinned-index

davishmcclurg committed:
More review feedback

- use same example `match` query in docs
- remove some unnecessary `this` references
- remove unnecessary constructors
- remove some single-line if statements

davishmcclurg committed:
Add alias test for pinned docs

This tests pinning the same document using different index aliases.
Previously, it would cause errors when scores were combined across
`should` clauses. Switching to `DisjunctionMaxQuery` solves the issue by
using the higher score instead.

jtibshirani committed:
Add REST yaml test, plus small changes

davishmcclurg commented:
Thanks for all the help @jtibshirani!
